In [1]:
# Open the jsonfile and display the first level of keys
# open API results WITH json module
import json
with open('example_yelp_results.json') as f:   #adjust for your path
    json_file = json.load(f)
    
## What type is top-level of json?
type(json_file)

dict

In [2]:
# what are the keys?
json_file.keys()

dict_keys(['businesses', 'total', 'region'])

We have three keys to explore.  We will start by exploring region.


# Region

In [3]:
# what type is region?
type(json_file['region'])

dict

In [4]:
# what keys are in region?
json_file['region'].keys()

dict_keys(['center'])

In [5]:
# what type is center?
type(json_file['region']['center'])

dict

In [6]:
# what are the keys in center?
json_file['region']['center'].keys()

dict_keys(['longitude', 'latitude'])

In [7]:
# Visualize the whole center dict, since lat and long are likely single values
json_file['region']['center']

{'longitude': -76.607666015625, 'latitude': 39.29617825404647}

# Total

In [8]:
# what is stored under the "total" key?
type(json_file['total'])

int

In [9]:
# What is the value?
json_file['total']

435

It looks like there were 435 total results for this Yelp search.

# Businesses

In [10]:
# What is in the businesses key?
type(json_file['businesses'])

list

In [11]:
# How long is businesses?
len(json_file['businesses'])

20

In [12]:
# There are 20 businesses stored in businesses
# What does the first entry look like?
json_file['businesses'][0]

{'id': 'D9A33FM394q99o4QtK5YwA',
 'alias': 'faidleys-seafood-baltimore-3',
 'name': 'Faidleys Seafood',
 'image_url': 'https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDCVS7pGopH6GZcfjBA/o.jpg',
 'is_closed': False,
 'url': 'https://www.yelp.com/biz/faidleys-seafood-baltimore-3?adjust_creative=KJtcufKUS887p24u6rjVIQ&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=KJtcufKUS887p24u6rjVIQ',
 'review_count': 1184,
 'categories': [{'alias': 'seafood', 'title': 'Seafood'},
  {'alias': 'beerbar', 'title': 'Beer Bar'}],
 'rating': 4.0,
 'coordinates': {'latitude': 39.291696, 'longitude': -76.62224},
 'transactions': ['delivery'],
 'price': '$$',
 'location': {'address1': '203 N Paca St',
  'address2': '',
  'address3': '',
  'city': 'Baltimore',
  'zip_code': '21201',
  'country': 'US',
  'state': 'MD',
  'display_address': ['203 N Paca St', 'Baltimore, MD 21201']},
 'phone': '+14107274898',
 'display_phone': '(410) 727-4898',
 'distance': 1349.560720156645}

In [14]:
# It is a dictionary, what are the keys?
json_file['businesses'][0].keys()

dict_keys(['id', 'alias', 'name', 'image_url', 'is_closed', 'url', 'review_count', 'categories', 'rating', 'coordinates', 'transactions', 'price', 'location', 'phone', 'display_phone', 'distance'])

# Records

There is a name for a particular type of json format in which we have a list that contains dictionaries with matching keys for each item.

We call these "records".

The "businesses" key in our json data is a list of records.

Records can be turned into DataFrames for easier inspection!

In [16]:
import pandas as pd
df_businesses = pd.DataFrame(json_file['businesses'])
df_businesses.head()

id                                 alias  \
0  D9A33FM394q99o4QtK5YwA          faidleys-seafood-baltimore-3   
1  u65W69AhbjUlvJJBkEhGNQ        miss-shirleys-cafe-baltimore-9   
2  ieS_5zqxDHcWMCm8BKUYbg  thames-street-oyster-house-baltimore   
3  6am8TZAFnvND52MOz-Yctg     mamas-on-the-half-shell-baltimore   
4  ISn7WyGQaFpsVSRSh0NSqg                sal-and-sons-baltimore   

                         name  \
0            Faidleys Seafood   
1         Miss Shirley's Cafe   
2  Thames Street Oyster House   
3    Mama's On The Half Shell   
4                Sal and Sons   

                                           image_url  is_closed  \
0  https://s3-media3.fl.yelpcdn.com/bphoto/OTjVDC...      False   
1  https://s3-media4.fl.yelpcdn.com/bphoto/9FsOyV...      False   
2  https://s3-media1.fl.yelpcdn.com/bphoto/9hGjo5...      False   
3  https://s3-media2.fl.yelpcdn.com/bphoto/HWY8OF...      False   
4  https://s3-media3.fl.yelpcdn.com/bphoto/LmVL4j...      False   

                                                 url  review_count  \
0  https://www.yelp.com/biz/faidleys-seafood-balt...          1184   
1  https://www.yelp.com/biz/miss-shirleys-cafe-ba...          2919   
2  https://www.yelp.com/biz/thames-street-oyster-...          2729   
3  https://www.yelp.com/biz/mamas-on-the-half-she...          1279   
4  https://www.yelp.com/biz/sal-and-sons-baltimor...           153   

                                          categories  rating  \
0  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.0   
1  [{'alias': 'breakfast_brunch', 'title': 'Break...     4.0   
2  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   
3  [{'alias': 'bars', 'title': 'Bars'}, {'alias':...     4.0   
4  [{'alias': 'seafood', 'title': 'Seafood'}, {'a...     4.5   

                                         coordinates        transactions  \
0    {'latitude': 39.291696, 'longitude': -76.62224}          [delivery]   
1  {'latitude': 39.2870995, 'longitude': -76.6053...  [delivery, pickup]   
2     {'latitude': 39.28214, 'longitude': -76.59162}          [delivery]   
3   {'latitude': 39.27986, 'longitude': -76.5752399}  [delivery, pickup]   
4       {'latitude': 39.284, 'longitude': -76.59337}          [delivery]   

  price                                           location         phone  \
0    $$  {'address1': '203 N Paca St', 'address2': '', ...  +14107274898   
1    $$  {'address1': '750 E Pratt St', 'address2': '',...  +14105285373   
2   $$$  {'address1': '1728 Thames St', 'address2': '',...  +14434497726   
3    $$  {'address1': '2901 Odonnell St', 'address2': '...  +14102763160   
4     $  {'address1': '1640 Aliceanna St', 'address2': ...  +14106751466   

    display_phone     distance  
0  (410) 727-4898  1349.560720  
1  (410) 528-5373  1028.736468  
2  (443) 449-7726  2090.712792  
3  (410) 276-3160  3328.825798  
4  (410) 675-1466  1817.406978

Even if we didn't save this dataframe as a variable, temporarily converting lists in JSON files to DataFrames can be very helpful for sifting through the file contents!

Now, the DataFrame definitely is not perfectly formatted. Take note of the "categories", "coordinates", and "location" columns.

While making our records into a DataFrame helped us organize and digest the business data, we will need to do additional cleaning on the columns that are filled with dictionaries or lists.

It is shown how to deal with these scenarios on " Advanced Transformations with Pandas ".